Panic when registry unreachable in retry handling

### Terraform Version
```
Terraform v0.12.23
```

### Terraform Configuration Files
Does not seems to be related to HCL, as the crash happened while initizing the `aws` provider:
```hcl
provider "aws" {
  region  = module.regional-common.region
  profile = "corp-example-com"
}
```

### Debug Output
```
Initializing provider plugins...
- Checking for available provider plugins...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1905347]

goroutine 1 [running]:
github.com/hashicorp/terraform/registry.maxRetryErrorHandler(0x0, 0x2ea8a80, 0xc000658a20, 0x2, 0x3, 0x3, 0x2ea8a80)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/registry/client.go:422 +0xe7
github.com/hashicorp/go-retryablehttp.(*Client).Do(0xc0000944e0, 0xc00064b260, 0x2a53998, 0x13, 0xc00049ece8)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/pkg/mod/github.com/hashicorp/go-retryablehttp@v0.5.2/client.go:452 +0x69a
github.com/hashicorp/terraform/registry.(*Client).TerraformProviderVersions(0xc0003cd080, 0xc0006c3db0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/registry/client.go:305 +0x46b
github.com/hashicorp/terraform/plugin/discovery.(*ProviderInstaller).listProviderVersions(0xc000294770, 0xc00003edb6, 0x3, 0x2a3a06c, 0x1, 0x2a584ad, 0x15, 0x0, 0x102d051, 0x203000)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/plugin/discovery/get.go:477 +0x7e
github.com/hashicorp/terraform/plugin/discovery.(*ProviderInstaller).Get(0xc000294770, 0xc00003edb6, 0x3, 0x2a3a06c, 0x1, 0x2a584ad, 0x15, 0x3e6dba0, 0x0, 0x0, ...)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/plugin/discovery/get.go:125 +0xde
github.com/hashicorp/terraform/command.(*InitCommand).getProviders(0xc0004a3a40, 0xc0001befc0, 0x0, 0x0, 0xc000470700, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/command/init.go:530 +0x4c3
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc0004a3a40, 0xc00000c090, 0x0, 0x0, 0xc00008b6c0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/command/init.go:347 +0xfc4
github.com/mitchellh/cli.(*CLI).Run(0xc00034a280, 0xc00034a280, 0xc000545cc0, 0x1)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x1da
main.wrappedMain(0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/main.go:239 +0xc43
main.realMain(0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/main.go:103 +0xb4
main.main()
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/main.go:39 +0x3a
```

### Crash Output

```
2020/03/16 17:21:40 [INFO] Terraform version: 0.12.23  
2020/03/16 17:21:40 [INFO] Go runtime version: go1.13.8
2020/03/16 17:21:40 [INFO] CLI args: []string{"/usr/local/bin/terraform", "init"}
2020/03/16 17:21:40 [DEBUG] Attempting to open CLI config file: /Users/tnn/.terraformrc
2020/03/16 17:21:40 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2020/03/16 17:21:40 [INFO] CLI command args: []string{"init"}
2020/03/16 17:21:40 [TRACE] ModuleInstaller: installing child modules for . into .terraform/modules
2020/03/16 17:21:40 [DEBUG] Module installer: begin environment-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: environment-common is not yet installed
2020/03/16 17:21:40 [TRACE] ModuleInstaller: cleaning directory .terraform/modules/environment-common prior to install of environment-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: environment-common has local path "../../environment-common"
2020/03/16 17:21:40 [TRACE] ModuleInstaller: environment-common uses directory from parent: ../../environment-common
2020/03/16 17:21:40 [DEBUG] Module installer: environment-common installed at ../../environment-common
2020/03/16 17:21:40 [DEBUG] Module installer: begin global-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: global-common is not yet installed
2020/03/16 17:21:40 [TRACE] ModuleInstaller: cleaning directory .terraform/modules/global-common prior to install of global-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: global-common has local path "../../../../global-common"
2020/03/16 17:21:40 [TRACE] ModuleInstaller: global-common uses directory from parent: ../../../../global-common
2020/03/16 17:21:40 [DEBUG] Module installer: global-common installed at ../../../../global-common
2020/03/16 17:21:40 [DEBUG] Module installer: begin regional-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: regional-common is not yet installed
2020/03/16 17:21:40 [TRACE] ModuleInstaller: cleaning directory .terraform/modules/regional-common prior to install of regional-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: regional-common has local path "../regional-common"
2020/03/16 17:21:40 [TRACE] ModuleInstaller: regional-common uses directory from parent: ../regional-common
2020/03/16 17:21:40 [DEBUG] Module installer: regional-common installed at ../regional-common
2020/03/16 17:21:40 [DEBUG] Module installer: begin site-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: site-common is not yet installed
2020/03/16 17:21:40 [TRACE] ModuleInstaller: cleaning directory .terraform/modules/site-common prior to install of site-common
2020/03/16 17:21:40 [TRACE] ModuleInstaller: site-common has local path "../../../site-common"
2020/03/16 17:21:40 [TRACE] ModuleInstaller: site-common uses directory from parent: ../../../site-common
2020/03/16 17:21:40 [DEBUG] Module installer: site-common installed at ../../../site-common
2020/03/16 17:21:40 [DEBUG] Module installer: begin vpc
2020/03/16 17:21:40 [TRACE] ModuleInstaller: vpc is not yet installed
2020/03/16 17:21:40 [TRACE] ModuleInstaller: cleaning directory .terraform/modules/vpc prior to install of vpc
2020/03/16 17:21:40 [TRACE] ModuleInstaller: vpc has local path "../../../../modules/aws/vpc"
2020/03/16 17:21:40 [TRACE] ModuleInstaller: vpc uses directory from parent: ../../../../modules/aws/vpc
2020/03/16 17:21:40 [DEBUG] Module installer: vpc installed at ../../../../modules/aws/vpc
2020/03/16 17:21:40 [TRACE] modsdir: writing modules manifest to .terraform/modules/modules.json
2020/03/16 17:21:40 [TRACE] Meta.Backend: no config given or present on disk, so returning nil config
2020/03/16 17:21:40 [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
2020/03/16 17:21:40 [DEBUG] New state was assigned lineage "52810a7f-fec0-54b1-331e-291a7f7ecf7d"
2020/03/16 17:21:40 [TRACE] Meta.Backend: using default local state only (no backend configuration, and no existing initialized backend)
2020/03/16 17:21:40 [TRACE] Meta.Backend: instantiated backend of type <nil>
2020/03/16 17:21:40 [DEBUG] checking for provider in "."
2020/03/16 17:21:40 [DEBUG] checking for provider in "/usr/local/bin"
2020/03/16 17:21:40 [DEBUG] checking for provisioner in "."
2020/03/16 17:21:40 [DEBUG] checking for provisioner in "/usr/local/bin"
2020/03/16 17:21:40 [INFO] Failed to read plugin lock file .terraform/plugins/darwin_amd64/lock.json: open .terraform/plugins/darwin_amd64/lock.json: no such file or directory
2020/03/16 17:21:40 [TRACE] Meta.Backend: backend <nil> does not support operations, so wrapping it in a local backend
2020/03/16 17:21:40 [TRACE] backend/local: state manager for workspace "default" will:
 - read initial snapshot from terraform.tfstate
 - write new snapshots to terraform.tfstate
 - create any backup at terraform.tfstate.backup
2020/03/16 17:21:40 [TRACE] statemgr.Filesystem: reading initial snapshot from terraform.tfstate
2020/03/16 17:21:40 [TRACE] statemgr.Filesystem: snapshot file has nil snapshot, but that's okay
2020/03/16 17:21:40 [TRACE] statemgr.Filesystem: read nil snapshot
2020/03/16 17:21:40 [DEBUG] checking for provider in "."
2020/03/16 17:21:40 [DEBUG] checking for provider in "/usr/local/bin"
2020/03/16 17:21:40 [DEBUG] plugin requirements: "aws"=""
2020/03/16 17:21:40 [DEBUG] Service discovery for registry.terraform.io at https://registry.terraform.io/.well-known/terraform.json
2020/03/16 17:21:40 [TRACE] HTTP client GET request to https://registry.terraform.io/.well-known/terraform.json
2020/03/16 17:21:43 [ERR] Checkpoint error: Get https://checkpoint-api.hashicorp.com/v1/check/terraform?arch=amd64&os=darwin&signature=3386d2ad-52b7-bc74-1ef9-e8692736a29f&version=0.12.23: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2020/03/16 17:21:43 [DEBUG] fetching provider versions from "https://registry.terraform.io/v1/providers/-/aws/versions"
2020/03/16 17:21:43 [TRACE] HTTP client GET request to https://registry.terraform.io/v1/providers/-/aws/versions
2020/03/16 17:21:54 [TRACE] HTTP client GET request to https://registry.terraform.io/v1/providers/-/aws/versions
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1905347]

goroutine 1 [running]:
github.com/hashicorp/terraform/registry.maxRetryErrorHandler(0x0, 0x2ea8a80, 0xc000658a20, 0x2, 0x3, 0x3, 0x2ea8a80)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/registry/client.go:422 +0xe7
github.com/hashicorp/go-retryablehttp.(*Client).Do(0xc0000944e0, 0xc00064b260, 0x2a53998, 0x13, 0xc00049ece8)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/pkg/mod/github.com/hashicorp/go-retryablehttp@v0.5.2/client.go:452 +0x69a
github.com/hashicorp/terraform/registry.(*Client).TerraformProviderVersions(0xc0003cd080, 0xc0006c3db0, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/registry/client.go:305 +0x46b
github.com/hashicorp/terraform/plugin/discovery.(*ProviderInstaller).listProviderVersions(0xc000294770, 0xc00003edb6, 0x3, 0x2a3a06c, 0x1, 0x2a584ad, 0x15, 0x0, 0x102d051, 0x203000)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/plugin/discovery/get.go:477 +0x7e
github.com/hashicorp/terraform/plugin/discovery.(*ProviderInstaller).Get(0xc000294770, 0xc00003edb6, 0x3, 0x2a3a06c, 0x1, 0x2a584ad, 0x15, 0x3e6dba0, 0x0, 0x0, ...)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/plugin/discovery/get.go:125 +0xde
github.com/hashicorp/terraform/command.(*InitCommand).getProviders(0xc0004a3a40, 0xc0001befc0, 0x0, 0x0, 0xc000470700, 0x0, 0x0, 0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/command/init.go:530 +0x4c3
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc0004a3a40, 0xc00000c090, 0x0, 0x0, 0xc00008b6c0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/command/init.go:347 +0xfc4
github.com/mitchellh/cli.(*CLI).Run(0xc00034a280, 0xc00034a280, 0xc000545cc0, 0x1)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x1da
main.wrappedMain(0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/main.go:239 +0xc43
main.realMain(0x0)
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/main.go:103 +0xb4
main.main()
	/private/tmp/terraform-20200306-88333-nvuhet/terraform-0.12.23/src/github.com/hashicorp/terraform/main.go:39 +0x3a
```
### Expected Behavior
Initialize the AWS provider and check/fetch latest provider plugin version. 

### Actual Behavior
While initalizing the provider plugin, terraform encountered a nill pointer dereference or invalid memory address.

### Steps to Reproduce
1. `terraform init`

### Additional Context
Running on Mac OS X from the command line via zsh. On the second attempt, it worked as expected and I have been unable to reproduce the crash.
### References

--- a/registry/client.go
+++ b/registry/client.go
@@ -14,7 +14,7 @@ import (
 	"time"
 
 	"github.com/hashicorp/go-retryablehttp"
-	"github.com/hashicorp/terraform-svchost"
+	svchost "github.com/hashicorp/terraform-svchost"
 	"github.com/hashicorp/terraform-svchost/disco"
 	"github.com/hashicorp/terraform/helper/logging"
 	"github.com/hashicorp/terraform/httpclient"
@@ -413,15 +413,23 @@ func maxRetryErrorHandler(resp *http.Response, err error, numTries int) (*http.R
 		resp.Body.Close()
 	}
 
+	// Additional error detail: if we have a response, use the status code;
+	// if we have an error, use that; otherwise nothing. We will never have
+	// both response and error.
 	var errMsg string
-	if err != nil {
-		errMsg = fmt.Sprintf(" %s", err)
+	if resp != nil {
+		errMsg = fmt.Sprintf(": %d", resp.StatusCode)
+	} else if err != nil {
+		errMsg = fmt.Sprintf(": %s", err)
 	}
+
+	// This function is always called with numTries=RetryMax+1. If we made any
+	// retry attempts, include that in the error message.
 	if numTries > 1 {
-		return resp, fmt.Errorf("the request failed after %d attempts, please try again later: %d%s",
-			numTries, resp.StatusCode, errMsg)
+		return resp, fmt.Errorf("the request failed after %d attempts, please try again later%s",
+			numTries, errMsg)
 	}
-	return resp, fmt.Errorf("the request failed, please try again later: %d%s", resp.StatusCode, errMsg)
+	return resp, fmt.Errorf("the request failed, please try again later%s", errMsg)
 }
 
 // configureRequestTimeout configures the registry client request timeout from

--- a/registry/client_test.go
+++ b/registry/client_test.go
@@ -314,6 +314,61 @@ func TestLookupModuleRetryError(t *testing.T) {
 	}
 }
 
+func TestLookupModuleNoRetryError(t *testing.T) {
+	// Disable retries
+	discoveryRetry = 0
+	defer configureDiscoveryRetry()
+
+	server := test.RegistryRetryableErrorsServer()
+	defer server.Close()
+
+	client := NewClient(test.Disco(server), nil)
+
+	src := "example.com/test-versions/name/provider"
+	modsrc, err := regsrc.ParseModuleSource(src)
+	if err != nil {
+		t.Fatal(err)
+	}
+	resp, err := client.ModuleVersions(modsrc)
+	if err == nil {
+		t.Fatal("expected request to fail", err)
+	}
+	if resp != nil {
+		t.Fatal("unexpected response", *resp)
+	}
+
+	// verify maxRetryErrorHandler handler returned the error
+	if !strings.Contains(err.Error(), "the request failed, please try again later") {
+		t.Fatal("unexpected error, got:", err)
+	}
+}
+
+func TestLookupModuleNetworkError(t *testing.T) {
+	server := test.RegistryRetryableErrorsServer()
+	client := NewClient(test.Disco(server), nil)
+
+	// Shut down the server to simulate network failure
+	server.Close()
+
+	src := "example.com/test-versions/name/provider"
+	modsrc, err := regsrc.ParseModuleSource(src)
+	if err != nil {
+		t.Fatal(err)
+	}
+	resp, err := client.ModuleVersions(modsrc)
+	if err == nil {
+		t.Fatal("expected request to fail", err)
+	}
+	if resp != nil {
+		t.Fatal("unexpected response", *resp)
+	}
+
+	// verify maxRetryErrorHandler handler returned the correct error
+	if !strings.Contains(err.Error(), "the request failed after 2 attempts, please try again later") {
+		t.Fatal("unexpected error, got:", err)
+	}
+}
+
 func TestLookupProviderVersions(t *testing.T) {
 	server := test.Registry()
 	defer server.Close()

--- a/registry/test/mock_registry.go
+++ b/registry/test/mock_registry.go
@@ -12,7 +12,7 @@ import (
 	"strings"
 
 	version "github.com/hashicorp/go-version"
-	"github.com/hashicorp/terraform-svchost"
+	svchost "github.com/hashicorp/terraform-svchost"
 	"github.com/hashicorp/terraform-svchost/auth"
 	"github.com/hashicorp/terraform-svchost/disco"
 	"github.com/hashicorp/terraform/httpclient"

